syntax = "proto3";

package product;

service ProductService {
  rpc GetProduct (GetProductRequest) returns (ProductResponse);
  rpc ListProducts (ListProductsRequest) returns (ListProductsResponse);
  rpc CreateProduct (CreateProductRequest) returns (ProductResponse);
  rpc SearchProduct (SearchProductRequest) returns (ListProductsResponse);
}

message Product {
  int32 id = 1;
  string name = 2;
  string description = 3;
  string category = 4;
  double price = 5;
  int32 inventory = 6;
}

message GetProductRequest {
  int32 id = 1;
}

message ListProductsRequest {
  int32 page = 1;
  int32 page_size = 2;
  string category = 3;
}

message CreateProductRequest {
  string name = 1;
  string description = 2;
  string category = 3;
  double price = 4;
}

message SearchProductRequest {
  string query = 1;
  int32 page = 2;
  int32 page_size = 3;
}

message ProductResponse {
  bool success = 1;
  string message = 2;
  Product product = 3;
}

message ListProductsResponse {
  repeated Product products = 1;
  int32 total = 2;
}

